Geometry/HGCalCommonData/plugins/DDHGCalEEFileAlgo.cc

```diff
@@ -0,0 +1,422 @@
+#endif
+      zz += hthick;
+    }  // End of loop over layers in a block
+    zi = zo;
+    laymin = laymax;
+    // Make consistency check of all the partitions of the block
+    if (std::abs(thickTot - layerThick_[i]) >= tol2_) {
+      if (thickTot > layerThick_[i]) {
+        edm::LogError("HGCalGeom") << "Thickness of the partition " << layerThick_[i] << " is smaller than " << thickTot
+                                   << ": thickness of all its components **** ERROR ****";
+      } else {
+        edm::LogWarning("HGCalGeom") << "Thickness of the partition " << layerThick_[i] << " does not match with "
+                                     << thickTot << " of the components";
+      }
+    }
+  }  // End of loop over blocks
+}
+
+void DDHGCalEEFileAlgo::positionSensitive(
+    const DDLogicalPart& glog, double rin, double rout, double zpos, int layertype, int layer, DDCompactView& cpv) {
+  static const double sqrt3 = std::sqrt(3.0);
+  int layercenter = layerCenter_[layer];
+  double r = 0.5 * (waferSize_ + waferSepar_);
+  double R = 2.0 * r / sqrt3;
+  double dy = 0.75 * R;
+  int N = (int)(0.5 * rout / r) + 2;
+  const auto& xyoff = geomTools_.shiftXY(layercenter, (waferSize_ + waferSepar_));
+#ifdef EDM_ML_DEBUG
+  int ium(0), ivm(0), iumAll(0), ivmAll(0), kount(0), ntot(0), nin(0);
+  std::vector<int> ntype(6, 0);
+  edm::LogVerbatim("HGCalGeom") << "DDHGCalEEFileAlgo: " << glog.ddname() << " rin:rout " << rin << ":" << rout
+                                << " zpos " << zpos << " N " << N << " for maximum u, v;  r " << r << " R " << R
+                                << " dy " << dy << " Shift " << xyoff.first << ":" << xyoff.second << " WaferSize "
+                                << (waferSize_ + waferSepar_);
+#endif
+  for (int u = -N; u <= N; ++u) {
+    int iu = std::abs(u);
+    for (int v = -N; v <= N; ++v) {
+      int iv = std::abs(v);
+      int nr = 2 * v;
+      int nc = -2 * u + v;
+      double xpos = xyoff.first + nc * r;
+      double ypos = xyoff.second + nr * dy;
+      const auto& corner = HGCalGeomTools::waferCorner(xpos, ypos, r, R, rin, rout, false);
+#ifdef EDM_ML_DEBUG
+      ++ntot;
+      if (((corner.first <= 0) && std::abs(u) < 5 && std::abs(v) < 5) || (std::abs(u) < 2 && std::abs(v) < 2)) {
+        edm::LogVerbatim("HGCalGeom") << "DDHGCalEEFileAlgo: " << glog.ddname() << " R " << rin << ":" << rout
+                                      << "\n Z " << zpos << " LayerType " << layertype << " u " << u << " v " << v
+                                      << " with " << corner.first << " corners";
+      }
+#endif
+      int type = HGCalWaferType::getType(HGCalWaferIndex::waferIndex(layer, u, v, false), waferIndex_, waferTypes_);
+      if (corner.first > 0 && type >= 0) {
+        int copy = type * 1000000 + iv * 100 + iu;
+        if (u < 0)
+          copy += 10000;
+        if (v < 0)
+          copy += 100000;
```

Comment:
When this code is migrated to DD4hep, these magic numbers (1e4, 1e5, 1e6) should be turned into constants.